futures-util/src/stream/stream/peek.rs

```diff
@@ -120,9 +125,51 @@ impl<S: Stream> Stream for Peekable<S> {
 // Forwarding impl of Sink from the underlying stream
 #[cfg(feature = "sink")]
 impl<S, Item> Sink<Item> for Peekable<S>
-    where S: Sink<Item> + Stream
+where
+    S: Sink<Item> + Stream,
 {
     type Error = S::Error;
 
     delegate_sink!(stream, Item);
 }
+
+/// Future for the [`peek()`] function from [`Peekable`]
+#[must_use = "futures do nothing unless polled"]
+pub struct Peek<'a, St: Stream> {
+    inner: Option<Pin<&'a mut Peekable<St>>>,
+}
+
+impl<St: Stream> Unpin for Peek<'_, St> {}
+
+impl<St: Stream> fmt::Debug for Peek<'_, St> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Peek")
+            .finish()
+    }
+}
+
+impl<St: Stream> FusedFuture for Peek<'_, St> {
+    fn is_terminated(&self) -> bool {
+        self.inner.is_none()
+    }
+}
+
+impl<'a, St> Future for Peek<'a, St>
+where
+    St: Stream,
+{
+    type Output = Option<&'a St::Item>;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if let Some(peekable) = self.inner.take() {
+            match peekable.poll_peek(cx) {
+                Either::Left(peekable) => {
+                    self.inner = Some(peekable);
+                    Poll::Pending
+                }
+                Either::Right(poll) => poll,
+            }
+        } else {
+            Poll::Pending
```

Comment:
I think this should be `panic!("Peek polled after completion")`, most futures panic after completion to (hopefully) make debugging extra polling easier.